adtl/transformations.py

```diff
@@ -171,3 +180,102 @@ def makeDateTime(date, time_24hr, date_format, timezone):
     tm = datetime.strptime(time_24hr, "%H:%M").time()
 
     return datetime.combine(dt, tm, tzinfo=zoneinfo.ZoneInfo(timezone)).isoformat()
+
+
+def splitDate(date, option: Literal["year", "month", "day"], format="%Y-%m-%d"):
+    "Splits a date into year, month, day"
+
+    if date in [None, ""]:
+        return None
+
+    sd = correctOldDate(date, format, return_datetime=True)
+
+    if sd is None:
+        return None
+
+    if option == "year":
+        return sd.year
+    elif option == "month":
+        return sd.month
+    elif option == "day":
+        return sd.day
+    else:
+        return None
+
+
+def startYear(currentdate, duration, dateformat="%Y-%m-%d", duration_type="years"):
+    """
+    Use to calculate year e.g. of birth from date (e.g. current date) and
+    duration (e.g. age)
+    """
+    if currentdate in [None, ""] or duration in [None, ""]:
+        return None
+
+    cd = correctOldDate(currentdate, dateformat, return_datetime=True)
+
+    if cd is None:
+        return None
+
+    if duration_type == "years":
+        return cd.year - float(duration)
+
+    elif duration_type == "months":
+        new_date = cd - relativedelta(months=floor(float(duration)))  # rounds down
+        return new_date.year
+
+    elif duration_type == "days":
+        new_date = cd - timedelta(days=float(duration))
+        return new_date.year
+
+
+def startMonth(currentdate, duration, dateformat="%Y-%m-%d", duration_type="months"):
+    """
+    Use to calculate month e.g. of birth from date (e.g. current date) and
+    duration (e.g. age)
+    """
+    if currentdate in [None, ""] or duration in [None, ""]:
+        return None
+
+    cd = correctOldDate(currentdate, dateformat, return_datetime=True)
+
+    if cd is None:
+        return None
+
+    if duration_type == "months":
+        new_date = cd - relativedelta(months=floor(float(duration)))
+        return new_date.month
+
+    elif duration_type == "days":
+        new_date = cd - timedelta(days=float(duration))
+        return new_date.month
+
+
+def correctOldDate(date, format, return_datetime=False):
+    """
+    the time module converts 2 digit dates as:
+    values 69-99 are mapped to 1969-1999, and values 0-68 are mapped to
+    2000-2068. This doesn't work for e.g. birthdates here where they are
+    frequently below the cutoff.
+
+    Switches the pivot point so that 22+ converts to 19xx.
+
+    Only use for birth dates to avoid unintentional conversion for recent
+    dates.
+    """
+
+    if date in [None, ""]:
+        return None
+
+    try:
+        cd = datetime.strptime(date, format)
+    except ValueError:
+        logging.error(f"Could not convert date {date!r} from date format {format!r}")
+        return None
+
+    if cd.year >= 2022 and "y" in format:
```

Comment:
I'd make this a non-optional parameter `epoch`. Best to be explicitly specified as 2022 within parsers.